Staging (#1928)
OpenLiberty/blogs: posts/2021-11-26-jakarta-ee-9.1.adoc

</httpEndpoint>
----

Configuring for all HTTP endpoints:

[source, xml]
----
<httpEndpoint id="defaultHttpEndpoint"
                  httpPort="9080"
		  httpsPort="9443"
                  headersRef="myHeadersID">
</httpEndpoint>
 
<httpEndpoint id="otherHttpEndpoint"
                  httpPort="9081"
                  httpsPort="9444"
                  headersRef="myHeadersID">
</httpEndpoint> 
     
<headers id="myHeadersID">
    <add>foo:bar</add>
    <add>foo:bar2</add>
    <set>customHeader:customValue</set>
    <setIfMissing>X-Forwarded-Proto:https</setIfMissing>
    <remove>Via</remove>
</headers>  
----

The `add` attribute allows multiple headers with the same name to be added to a response, similar to the HttpServletResponse's `addHeader` API. Similarly, the `set` attribute is analogous to the `setHeader` API, which sets a response header to the given name and value. This overwrites existing headers that share the same name. The `setIfMissing` attribute will only set the configured headers if they are not already present on the response. Lastly, the `remove` attribute will remove any response headers whose name matches a name defined by the configuration. 

> Each header entry for the `add`, `set`, and `setIfMissing` attributes can be provided as a standalone header name. Optionally, a header value can be added by appending the colon `:` character after every header name. Note, however, that the `remove` attribute only expects header names and not a header name:value pair.

Reviewer: standalone-->stand-alone

Author: ```suggestion
Each header entry for the `add`, `set`, and `setIfMissing` attributes can be provided as a stand-alone header name. Optionally, a header value can be added by appending the colon `:` character after every header name. Note, however, that the `remove` attribute only expects header names and not a header name:value pair.
```

+
and the following FFDC will be seen:
+   
[source, no_copy]
----
Exception = java.lang.RuntimeException
Source = com.ibm.ws.tcpchannel.internal.NBAcceptChannelSelector
probeid = 254
Stack Dump = java.lang.RuntimeException: java.nio.channels.NotYetBoundException
	at com.ibm.ws.tcpchannel.internal.NBAcceptChannelSelector.performRequest(NBAcceptChannelSelector.java:358)
	at com.ibm.ws.tcpchannel.internal.ChannelSelector.run(ChannelSelector.java:257)
	at java.base/java.lang.Thread.run(Thread.java:866)
Caused by: java.nio.channels.NotYetBoundException
	at java.base/sun.nio.ch.ServerSocketChannelImpl.begin(ServerSocketChannelImpl.java:246)
	at java.base/sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:283)
	at com.ibm.ws.tcpchannel.internal.NBAcceptChannelSelector.performRequest(NBAcceptChannelSelector.java:278)
	... 2 more
----
+
This issue has now been fixed so that all configured ports should start, or if there is a problem some meaningful error message should be logged.

* link:https://github.com/OpenLiberty/open-liberty/issues/18992[Application fails to restart in server.xml update scenario]
+

> We discovered an issue where an application would fail to restart, due to a race condition during server reconfiguration when multiple apps are installed. The problem occurs when one app starts before another app is finished uninstalling. In theory this shouldn't be a problem - however for this scenario these apps are sharing a VirtualHost configuration object, and in this case one app updates the parent VirtualHost as part of its uninstall process in such a way that the other gets into an invalid state. The server log will show an warning such as `CWWKZ0020I: Application <app_name> not updated`. This issue was fixed by fixing the race condition that caused the failure.

Reviewer: The server log will show an warning
--->
The server log will show a warning

Author: ```suggestion
We discovered an issue where an application would fail to restart, due to a race condition during server reconfiguration when multiple apps are installed. The problem occurs when one app starts before another app is finished uninstalling. In theory this shouldn't be a problem - however for this scenario these apps are sharing a VirtualHost configuration object, and in this case one app updates the parent VirtualHost as part of its uninstall process in such a way that the other gets into an invalid state. The server log will show a warning such as `CWWKZ0020I: Application <app_name> not updated`. This issue was fixed by fixing the race condition that caused the failure.
```

* link:https://github.com/OpenLiberty/open-liberty/issues/17599[HTTP upgrade to WebSocket can cause quiesce errors]
+

> When a websocket connection is started, it starts as an HTTP connection. Previously, If an error occured during the transition between an HTTP and a websocket connection, which was known to be a websocket upgrade, the error processing would neglect to decrement a connection counter, which then caused the server to believe there is an open connection during server stop.

Reviewer: occured--> occurred

> When a websocket connection is started, it starts as an HTTP connection. Previously, If an error occured during the transition between an HTTP and a websocket connection, which was known to be a websocket upgrade, the error processing would neglect to decrement a connection counter, which then caused the server to believe there is an open connection during server stop.

Reviewer: websocket--> WebSocket (3 instances)

Author: ```suggestion
When a WebSocket connection is started, it starts as an HTTP connection. Previously, If an error occurred during the transition between an HTTP and a WebSocket connection, which was known to be a WebSocket upgrade, the error processing would neglect to decrement a connection counter, which then caused the server to believe there is an open connection during server stop.
```

There were two scenarios where these quiesce errors would occur:
+

> ** When a read error occured during the transition between an HTTP and a websocket connection, the error processing neglected to decrement a connection counter, which then causes the server to believe there is an open connection during server stop.

Reviewer: occured--> occurred

> ** When a read error occured during the transition between an HTTP and a websocket connection, the error processing neglected to decrement a connection counter, which then causes the server to believe there is an open connection during server stop.

Reviewer: websocket--> WebSocket

Author: ```suggestion
** When a read error occurred during the transition between an HTTP and a WebSocket connection, the error processing neglected to decrement a connection counter, which then causes the server to believe there is an open connection during server stop.
```

+

> ** If a client immediately closed the websocket connection after it was opened, the original upgrade request handling may not have had enough time to close properly on the server. Once again, the connection counter failed to decrement leading the server to believe there is an open connection during the server stop.

Reviewer: websocket--> WebSocket

Author: ```suggestion
** If a client immediately closed the WebSocket connection after it was opened, the original upgrade request handling may not have had enough time to close properly on the server. Once again, the connection counter failed to decrement leading the server to believe there is an open connection during the server stop.
```

+
This issue has been fixed by adding a new flag called `decrementNeeded` which helps to ensure that the decrement is not neglected.

* link:https://github.com/OpenLiberty/open-liberty/issues/19227[Ensure ServletRequestListener#requestDestroyed is always called]
+
We discovered a bug where the `ServletRequestListener#requestDestroyed` call does not occur, if an exception occurs during async servlet while an `appSecurity-x.0` is enabled. For this bug to occur, two conditions must be met: the `webContainer` property `deferServletRequestListenerDestroyOnError` is true and an `appSecurity-x.0` feature is enabled. This issue has now been resolved.

* link:https://github.com/OpenLiberty/open-liberty/issues/19197[ClassCastException in JSP relating to JDT internal classes]
+

> Open Liberty 21.0.0.9 introduced a bug where the following error occured for certain class lookups in JSP:

Reviewer: occured--> occurred

Author: ```suggestion
Open Liberty 21.0.0.9 introduced a bug where the following error occurred for certain class lookups in JSP:
```
